prototype2/src/bin.rs

```diff
@@ -1,44 +1,40 @@
 extern crate prototype2;
+extern crate clap;
 
-use std::env;
 use std::str::FromStr;
 use std::net::ToSocketAddrs;
+use clap::{App, Arg, ArgMatches, SubCommand};
+use clap::AppSettings::SubcommandRequired;
 
 static EXAMPLE_SERVER_COMMAND: &'static str = "space_coop server 8888";
 static EXAMPLE_CLIENT_COMMAND: &'static str = "space_coop client 9999 192.168.0.1:8888";
 
 fn main() {
-  let app_type = env::args().nth(1);
+  let matches = App::new("space coop")
+    .usage(format!("\t{}\n\t{}", EXAMPLE_SERVER_COMMAND, EXAMPLE_CLIENT_COMMAND).as_ref())
+    .settings(&[SubcommandRequired])
+    .subcommand(SubCommand::with_name("server")
+      .usage(EXAMPLE_SERVER_COMMAND)
+      .arg(Arg::with_name("port").required(true)))
+    .subcommand(SubCommand::with_name("client")
+      .usage(EXAMPLE_CLIENT_COMMAND)
+      .arg(Arg::with_name("port").required(true))
+      .arg(Arg::with_name("server address").required(true)))
+    .get_matches();
 
-  match app_type.as_ref().map(|v| v.as_ref()) {
-    Some("server") => {
-      let port_opt = env::args().nth(2).and_then(|v| u16::from_str(&v).ok());
-      match port_opt {
-        Some(port) => prototype2::server::start(port),
-        None => {
-          println!("ERROR: Specify a port, as in \"{}\"",
-                   EXAMPLE_SERVER_COMMAND)
-        },
-      }
-    },
-    Some("client") => {
-      let port_opt = env::args().nth(2).and_then(|v| u16::from_str(&v).ok());
-      let socket_addr_opt = env::args()
-        .nth(3)
-        .and_then(|v| v.to_socket_addrs().ok())
-        .and_then(|mut socket_addr_iter| socket_addr_iter.next());
-      match (port_opt, socket_addr_opt) {
-        (Some(port), Some(addr)) => prototype2::client::start(port, addr),
-        _ => {
-          println!("ERROR: Specify a local port and a server ip and port, as in \"{}\"",
-                   EXAMPLE_CLIENT_COMMAND)
-        },
-      }
-    },
-    _ => {
-      println!("ERROR: Specify an application type from [server, client], as in, \"{}\", or \"{}\"",
-               EXAMPLE_SERVER_COMMAND,
-               EXAMPLE_CLIENT_COMMAND)
-    },
+  if let Some(server_matches) = matches.subcommand_matches("server") {
```

Comment:
What happens when this doesn't match at all? Does it just exit? Thats pretty unintuitive.